### PR TITLE
ISSUE #2978, Add system.query_log

### PR DESCRIPTION
**Pull request by SGZW**
I hereby agree to the terms of the CLA available at: https://databend.rs/policies/cla/

## Summary

solve #2978 

## Changelog

- New Feature

## Test Plan

Unit Tests

Stateless Tests

### CODE REVIEW

**Comment by SGZW** [comment]
I'm sorry that this task was completed late

**Comment by BohuTANG** [comment]
> I'm sorry that this task was completed late

Thanks @SGZW , this looks great.
This almost ready, unit test needs some fix :D

File: query/src/configs/config_query.rs [hunk 1]
```diff
@@ -206,6 +207,14 @@ pub struct QueryConfig {
...
     #[serde(default)]
     pub wait_timeout_mills: u64,
+
+    #[structopt(
+    long,
+    env = MAX_QUERY_LOG_SIZE,
+    default_value = "10000"
+    )]
+    #[serde(default)]
+    pub max_query_log_size: usize,
```
Comment (BohuTANG) [inline query/src/configs/config_query.rs:217]: :+1:

**Comment by sundy-li** [comment]
How about using fuse engine to replace memory engine? I think it's better to persist in the query logs.

**Comment by SGZW** [comment]
> How about using fuse engine to replace memory engine? I think it's better to persist in the query logs.

it means multi query process can share query_log by fuse table?

**Comment by SGZW** [comment]
> How about using fuse engine to replace memory engine? I think it's better to persist in the query logs.

I think we can learn from query_log's implement of clickhouse which pesist query log in merge tree Periodic ？ I can try this in another pr

**Comment by BohuTANG** [comment]
> How about using fuse engine to replace memory engine? I think it's better to persist in the query logs.

We will add LRU to query_log later(The max rows is N, N default like 100000), memory engine is OK

**Comment by SGZW** [comment]
> > How about using fuse engine to replace memory engine? I think it's better to persist in the query logs.
> 
> We will add LRU to query_log later(The max rows is N, N default like 100000), memory engine is OK

what is LRU mean? fifo is

**Comment by BohuTANG** [comment]
> > > How about using fuse engine to replace memory engine? I think it's better to persist in the query logs.
> > 
> > 
> > We will add LRU to query_log later(The max rows is N, N default like 100000), memory engine is OK
> 
> what is LRU mean? FIFO isn't enough?

Sorry, FIFO.
It have implement in this patch

**Pull request merged**